### ISSUE TITLE
[FP] Statement could be decoupled from foreach with output

### ISSUE THREAD

**Issue by orklah** [kalessil/phpinspectionsea#1154]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Issue type     | False-Positive                       |
| Plugin         | Php Inspections (EA Extended)|

Current behaviour (description/screenshot:)
--
When there is a statement used to output datas, it shouldn't be considered as decoupled from the loop.
I'm looking at a code where foreach loops are used to construct a template. There is no way the "echo $var" statement could be moved out of the foreach because the template wouldn't look the same at all.

Expected behaviour
--
The most we could do is check that there isn't additional operations like:
"echo $var == 12" could still suggest to move the "$var == 12" out of the loop, but when the statement is just "echo $var_equals_12", there is not much to do...

**Comment by kalessil** [comment]
@orklah, I need a real-life code snippet here, as this specific inspection is quite complex.

**Comment by orklah** [comment]
Here are two examples from wordpress source code:

![Example1](https://user-images.githubusercontent.com/9605520/50560697-b4bf2700-0d03-11e9-9155-cd44831eee68.png)
In this first one, there is a variable named $describedby_attr. It is actually defined before the loop with this pre-calculed html attributes:
```php
$describedby_attr = ( ! empty( $this->description ) ) ? ' aria-describedby="' . esc_attr( $description_id ) . '" ' : '';
```
The variable is then used as an output to render the full html markup.

The second one is maybe more clear:
![Example2](https://user-images.githubusercontent.com/9605520/50560796-85f58080-0d04-11e9-9b42-aeebe5071d53.PNG)
The goal is to generate a list of radio buttons. They all have differents values but the same name. The name is calculated out of the loop and then used as an echo inside the loop.
For this example, this could have been a smart move to use the function esc_attr() out of the loop since its behavior don't seem to depend on the loop context.

For both those example, I see no way to achieve the same result without using a variable defined out of the loop. The simple fact that there is an output (echo, print, ...) in the statement make this inspection irrelevant for this statement. You can't move an echo out of the loop and expect the result to be the same.

There were 9 similar cases in the wordpress source code when I launched the inspection against.

**Comment by kalessil** [comment]
Oh, got now, thank you. The inspection will skip loops in templating now.

**Comment by orklah** [comment]
It resolved some 5 of the 9 false positives I saw.

Here are some example of what stayed highlighted:
echo $var->something in templating
![](https://user-images.githubusercontent.com/9605520/50605260-1b8a3f00-0ec2-11e9-8636-c4c227f5cebb.PNG)
This case look like the previous ones very much, I don't understand why it didn't get resolved:
![](https://user-images.githubusercontent.com/9605520/50605261-1b8a3f00-0ec2-11e9-8e97-fd9d18067650.PNG)

Is there something that could be fixed?

**Comment by kalessil** [comment]
Yep, fixed those cases as well
